File: examples/payment-service-java/src/RetryHelper.java
package com.example.payment;

import java.util.concurrent.Callable;

public class RetryHelper {
    private final int maxAttempts;
    private final long delayMs;

    public RetryHelper(int maxAttempts, long delayMs) {
        this.maxAttempts = maxAttempts;
        this.delayMs = delayMs;
    }

    public <T> T execute(Callable<T> fn) throws Exception {
        int attempts = 0;
        Exception lastException = null;

        while (attempts < maxAttempts) {
            try {
                return fn.call();
            } catch (Exception e) {
                lastException = e;
                attempts++;
                if (attempts < maxAttempts) {
                    Thread.sleep(delayMs);
                }
            }
        }

        throw new RuntimeException(
            "Payment failed after " + maxAttempts + " attempts",
            lastException
        );
    }
}

File: examples/payment-service-kotlin/RetryHelper.kt
package com.example.payment

import kotlinx.coroutines.delay

data class RetryConfig(val maxAttempts: Int, val delayMs: Long)

suspend fun <T> withRetry(config: RetryConfig, fn: suspend () -> T): T {
    var attempts = 0
    var lastException: Exception? = null

    while (attempts < config.maxAttempts) {
        try {
            return fn()
        } catch (e: Exception) {
            lastException = e
            attempts++
            if (attempts < config.maxAttempts) {
                delay(config.delayMs)
            }
        }
    }

    throw RuntimeException("Payment failed after ${config.maxAttempts} attempts", lastException)
}
